File: sliding-window/904-Fruit-Into-Baskets.go
func totalFruit(fruits []int) int {
    type_fruit := make(map[int]int)
    res, left := 0, 0
    for right, _ := range fruits {
        if _, found := type_fruit[fruits[right]]; found {
            type_fruit[fruits[right]]++
        } else {
            type_fruit[fruits[right]] = 1
        }

        if len(type_fruit) > 2 {
            type_fruit[fruits[left]]--

            if type_fruit[fruits[left]] == 0 {
                delete(type_fruit, fruits[left])
            }
            left++
        }

        res = max(res, right - left + 1)
    }
    return res
}

func max(a int, b int) int {
    if a > b {
        return a
    }
    return b
}

File: reference/sliding-window/904-Fruit-Into-Baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        trees = {}
        left, max_fruits = 0, 0

        for right, fruit in enumerate(fruits):
            if fruit not in trees:
                trees[fruit] = 1
            else:
                trees[fruit] += 1

            while len(trees) > 2:
                trees[fruits[left]] -= 1
                if trees[fruits[left]] == 0:
                    del trees[fruits[left]]
                left += 1

            max_fruits = max(max_fruits, right - left + 1)

        return max_fruits